numpy/core/numeric.py

```diff
@@ -1691,19 +1691,14 @@ def indices(dimensions, dtype=int, sparse=False):
     dimensions = tuple(dimensions)
     N = len(dimensions)
     shape = (1,)*N
+    indices = tuple(arange(dim, dtype=dtype).reshape(
+                shape[:i] + (dim,) + shape[i+1:]) for i, dim in enumerate(dimensions))
     if sparse:
-        res = tuple()
+        return indices
     else:
-        res = empty((N,)+dimensions, dtype=dtype)
-    for i, dim in enumerate(dimensions):
-        idx = arange(dim, dtype=dtype).reshape(
-            shape[:i] + (dim,) + shape[i+1:]
-        )
-        if sparse:
-            res = res + (idx,)
-        else:
-            res[i] = idx
-    return res
+        if indices == ():
+            return np.array([])
```

Comment:
dtype is wrong.

I'd prefer to avoid special casing for exactly this reason. I think I prefered this as it was, as it was obviously correct for 0d arrays

Reply:
You're perfectly right. Reverting that change, but leaving the test anyway.

Reply:
Oops, I missed the further comments before commenting. I don't think it matters much either way, though I think I'd slightly prefer my option 3.

Reply:
Option 3 looks fine, but I'm not sure it's worth changing. It has a slightly larger memory footprint than this version, but that probably doesn't matter.